File: 1001-1500/1469_getLonelyNodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def getLonelyNodes(self, root: Optional[TreeNode]) -> List[int]:
        self.lonelyNodes = []
        self.dfs(root)
        return self.lonelyNodes
        
    def dfs(self, node):
        if not node:
            return
        if node.left and not node.right:
            self.lonelyNodes.append(node.left.val)
        if node.right and not node.left:
            self.lonelyNodes.append(node.right.val)
        self.dfs(node.left)
        self.dfs(node.right)
        return node

File: 1001-1500/1469_getLonelyNodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> getLonelyNodes(TreeNode root) {
        List<Integer> lonelyNodes = new ArrayList<>();
        dfs(root, lonelyNodes);
        return lonelyNodes;
        
    }
    
    public void dfs(TreeNode root, List<Integer> nodes) {
        if (root == null) {
            return;
        }
        
        if ((root.left != null && root.right == null) || (root.right != null && root.left == null)) {
            nodes.add((root.left != null) ? root.left.val : root.right.val);
        }
        
        dfs(root.left, nodes);
        dfs(root.right, nodes);
    }
}
